File: app/src/main/java/com/jobtick/activities/ViewAllOffersActivity.java
package com.jobtick.activities;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.material.appbar.MaterialToolbar;
import com.jobtick.R;
import com.jobtick.adapers.OfferListAdapter;
import com.jobtick.models.OfferModel;
import com.jobtick.pagination.PaginationListener;
import com.jobtick.utils.Constant;
import com.jobtick.utils.ConstantKey;
import com.jobtick.utils.Helper;
import com.jobtick.utils.SessionManager;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import butterknife.BindView;
import butterknife.ButterKnife;

import static com.jobtick.pagination.PaginationListener.PAGE_START;

public class ViewAllOffersActivity extends ActivityBase implements SwipeRefreshLayout.OnRefreshListener, OfferListAdapter.OnItemClickListener {

    @BindView(R.id.toolbar)
    MaterialToolbar toolbar;
    @BindView(R.id.recycler_view_all_ofers)
    RecyclerView recyclerViewAllOfers;
    @BindView(R.id.swipeRefresh)
    SwipeRefreshLayout swipeRefresh;

    private int currentPage = PAGE_START;
    private boolean isLastPage = false;
    private int totalPage = 10;
    private boolean isLoading = false;
    private OfferListAdapter offerListAdapter;
    private String str_slug;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_view_all_offers);
        ButterKnife.bind(this);

        Bundle bundle = getIntent().getExtras();
        if(bundle != null && bundle.getString(ConstantKey.SLUG) != null){
            str_slug = bundle.getString(ConstantKey.SLUG);
        }
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onBackPressed();
            }
        });
        initCOmponent();
    }

    private void initCOmponent() {
        recyclerViewAllOfers.setLayoutManager(new LinearLayoutManager(ViewAllOffersActivity.this, LinearLayoutManager.VERTICAL, false));
        recyclerViewAllOfers.setHasFixedSize(true);
        sessionManager = new SessionManager(ViewAllOffersActivity.this);
        offerListAdapter = new OfferListAdapter(ViewAllOffersActivity.this,false,new ArrayList<>());
        recyclerViewAllOfers.setAdapter(offerListAdapter);
        swipeRefresh.setOnRefreshListener(this);

        // use a linear layout manager
        LinearLayoutManager layoutManager = new LinearLayoutManager(ViewAllOffersActivity.this);
        offerListAdapter.setOnItemClickListener(this);
        swipeRefresh.setRefreshing(true);
        doApiCall();
        /*
         * add scroll listener while user reach in bottom load more will call
         */
        recyclerViewAllOfers.addOnScrollListener(new PaginationListener(layoutManager) {
            @Override
            protected void loadMoreItems() {
                isLoading = true;
                currentPage++;
                doApiCall();
            }

            @Override
            public boolean isLastPage() {
                return isLastPage;
            }

            @Override
            public boolean isLoading() {
                return isLoading;
            }
        });
    }


    private void doApiCall() {

        ArrayList<OfferModel> items = new ArrayList<>();
        Helper.closeKeyboard(ViewAllOffersActivity.this);
        StringRequest stringRequest = new StringRequest(Request.Method.GET, Constant.URL_TASKS+"/"+ str_slug + "/offers?page=" + currentPage ,
                new Response.Listener<String>() {
                    @Override
                    public void onResponse(String response) {
                        Log.e("responce_url", response);
                        // categoryArrayList.clear();
                        try {
                            JSONObject jsonObject = new JSONObject(response);
                            Log.e("json", jsonObject.toString());
                            if (!jsonObject.has("data")) {
                                showToast("some went to wrong", ViewAllOffersActivity.this);
                                return;
                            }
                            JSONArray jsonArray_data = jsonObject.getJSONArray("data");
                            for (int i = 0; jsonArray_data.length() > i; i++) {
                                JSONObject jsonObject_offers = jsonArray_data.getJSONObject(i);
                                OfferModel offerModel = new OfferModel().getJsonToModel(jsonObject_offers);
                                items.add(offerModel);
                            }

                            if (jsonObject.has("meta") && !jsonObject.isNull("meta")) {
                                JSONObject jsonObject_meta = jsonObject.getJSONObject("meta");
                                totalPage = jsonObject_meta.getInt("last_page");
                                Constant.PAGE_SIZE = jsonObject_meta.getInt("per_page");
                            }

                            /*
                             *manage progress view
                             */
                            if (currentPage != PAGE_START)
                                offerListAdapter.removeLoading();
                            offerListAdapter.addItems(items);

                            swipeRefresh.setRefreshing(false);
                            // check weather is last page or not
                            if (currentPage < totalPage) {
                                offerListAdapter.addLoading();
                            } else {
                                isLastPage = true;
                            }
                            isLoading = false;
                        } catch (JSONException e) {
                            hideProgressDialog();
                            Log.e("EXCEPTION", String.valueOf(e));
                            e.printStackTrace();
                        }
                    }
                },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        swipeRefresh.setRefreshing(false);
                        errorHandle1(error.networkResponse);
                    }
                }) {
            @Override
            public Map<String, String> getHeaders() throws AuthFailureError {
                Map<String, String> map1 = new HashMap<String, String>();
                map1.put("Content-Type", "application/x-www-form-urlencoded");
                map1.put("Authorization", "Bearer " + sessionManager.getAccessToken());
                return map1;
            }
        };

        stringRequest.setRetryPolicy(new DefaultRetryPolicy(0, -1,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
        RequestQueue requestQueue = Volley.newRequestQueue(ViewAllOffersActivity.this);
        requestQueue.add(stringRequest);
        Log.e("url", stringRequest.getUrl());
    }

    @Override
    public void onRefresh() {
        currentPage = PAGE_START;
        isLastPage = false;
        offerListAdapter.clear();
        doApiCall();
    }

    @Override
    public void onItemOfferClick(View view, OfferModel obj, int position, String action) {
        if (action.equalsIgnoreCase("reply")) {
            Intent intent = new Intent(ViewAllOffersActivity.this, PublicChatActivity.class);
            Bundle bundle = new Bundle();
            bundle.putParcelable(ConstantKey.OFFER_LIST_MODEL, obj);
            intent.putExtras(bundle);
            startActivityForResult(intent, 20);
        }
    }
}

File: app/src/main/java/com/jobtick/android/activities/ViewAllOffersActivity.kt
package com.jobtick.android.activities

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import com.android.volley.*
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.android.material.appbar.MaterialToolbar
import com.jobtick.android.BuildConfig
import com.jobtick.android.R
import com.jobtick.android.adapers.OfferListAdapter
import com.jobtick.android.models.OfferModel
import com.jobtick.android.pagination.PaginationListener
import com.jobtick.android.utils.Constant
import com.jobtick.android.utils.ConstantKey
import com.jobtick.android.utils.Helper
import com.jobtick.android.utils.SessionManager
import org.json.JSONException
import org.json.JSONObject
import timber.log.Timber
import java.util.*

class ViewAllOffersActivity : ActivityBase(), OnRefreshListener, OfferListAdapter.OnItemClickListener {
    private lateinit var toolbar: MaterialToolbar
    private lateinit var recyclerViewAllOfers: RecyclerView
    private lateinit var swipeRefresh: SwipeRefreshLayout
    private var currentPage = PaginationListener.PAGE_START
    private var isLastPageItem = false
    private var totalPage = 10
    private var isLoadingItem = false
    private var offerListAdapter: OfferListAdapter? = null
    private var strSlug: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_view_all_offers)
        setIDs()
        val bundle = intent.extras
        if (bundle?.getString(ConstantKey.SLUG) != null) {
            strSlug = bundle.getString(ConstantKey.SLUG)
        }
        toolbar.setNavigationOnClickListener { onBackPressed() }
        initCOmponent()
    }

    private fun setIDs() {
        toolbar = findViewById(R.id.toolbar)
        recyclerViewAllOfers = findViewById(R.id.recycler_view_all_ofers)
        swipeRefresh = findViewById(R.id.swipeRefresh)
    }

    private fun initCOmponent() {
        recyclerViewAllOfers.layoutManager = LinearLayoutManager(this@ViewAllOffersActivity, LinearLayoutManager.VERTICAL, false)
        recyclerViewAllOfers.setHasFixedSize(true)
        sessionManager = SessionManager(this@ViewAllOffersActivity)
        offerListAdapter = OfferListAdapter(this@ViewAllOffersActivity, false, ArrayList())
        recyclerViewAllOfers.adapter = offerListAdapter
        swipeRefresh.setOnRefreshListener(this)

        val layoutManager = LinearLayoutManager(this@ViewAllOffersActivity)
        offerListAdapter!!.setOnItemClickListener(this)
        swipeRefresh.isRefreshing = true
        doApiCall()
        recyclerViewAllOfers.addOnScrollListener(object : PaginationListener(layoutManager) {
            override fun loadMoreItems() {
                isLoadingItem = true
                currentPage++
                doApiCall()
            }
            override val isLastPage: Boolean
                get() = isLastPageItem
            override val isLoading: Boolean
                get() = isLoadingItem
        })
    }

    private fun doApiCall() {
        val items = ArrayList<OfferModel>()
        Helper.closeKeyboard(this@ViewAllOffersActivity)
        val stringRequest: StringRequest = object : StringRequest(Method.GET, Constant.URL_TASKS + "/" + strSlug + "/offers?page=" + currentPage,
                Response.Listener { response ->
                    Timber.e(response)
                    // categoryArrayList.clear();
                    try {
                        val jsonObject = JSONObject(response)
                        Timber.e(jsonObject.toString())
                        if (!jsonObject.has("data")) {
                            showToast("some went to wrong", this@ViewAllOffersActivity)
                            return@Listener
                        }
                        val jsonArrayData = jsonObject.getJSONArray("data")
                        var i = 0
                        while (jsonArrayData.length() > i) {
                            val jsonObjectOffers = jsonArrayData.getJSONObject(i)
                            val offerModel = OfferModel().getJsonToModel(jsonObjectOffers)
                            items.add(offerModel)
                            i++
                        }
                        if (jsonObject.has("meta") && !jsonObject.isNull("meta")) {
                            val jsonObjectMeta = jsonObject.getJSONObject("meta")
                            totalPage = jsonObjectMeta.getInt("last_page")
                            Constant.PAGE_SIZE = jsonObjectMeta.getInt("per_page")
                        }

                        /*
                                       *manage progress view
                                       */if (currentPage != PaginationListener.PAGE_START) offerListAdapter!!.removeLoading()
                        offerListAdapter!!.addItems(items)
                        swipeRefresh.isRefreshing = false
                        // check weather is last page or not
                        if (currentPage < totalPage) {
                            offerListAdapter!!.addLoading()
                        } else {
                            isLastPageItem = true
                        }
                        isLoadingItem = false
                    } catch (e: JSONException) {
                        hideProgressDialog()
                        Timber.e(e.toString())
                        e.printStackTrace()
                    }
                },
                Response.ErrorListener { error ->
                    swipeRefresh.isRefreshing = false
                    errorHandle1(error.networkResponse)
                }) {
            @Throws(AuthFailureError::class)
            override fun getHeaders(): Map<String, String> {
                val map1: MutableMap<String, String> = HashMap()
                map1["Content-Type"] = "application/x-www-form-urlencoded"
                map1["Authorization"] = "Bearer " + sessionManager.accessToken
                map1["Version"] = BuildConfig.VERSION_CODE.toString()
                return map1
            }
        }
        stringRequest.retryPolicy = DefaultRetryPolicy(0, -1,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)
        val requestQueue = Volley.newRequestQueue(this@ViewAllOffersActivity)
        requestQueue.add(stringRequest)
        Timber.e(stringRequest.url)
    }

    override fun onRefresh() {
        currentPage = PaginationListener.PAGE_START
        isLastPageItem = false
        offerListAdapter!!.clear()
        doApiCall()
    }

    override fun onItemOfferClick(view: View, obj: OfferModel, position: Int, action: String) {
        if (action.equals("reply", ignoreCase = true)) {
            val intent = Intent(this@ViewAllOffersActivity, PublicChatActivity::class.java)
            val bundle = Bundle()
            bundle.putParcelable(ConstantKey.OFFER_LIST_MODEL, obj)
            intent.putExtras(bundle)
            startActivityForResult(intent, 20)
        }
    }
}